File: test/file-benchmark.mjs
#!/usr/bin/env node
/**
 * 文件渲染性能测试脚本
 * 动态测试 static 目录下所有 PDF 文件的转换性能（前 10 页）
 */

import path from 'path';
import fs from 'fs';
import { fileURLToPath } from 'url';
import { convert, getPageCount, isAvailable, getVersion, getThreadPoolStats, destroyThreadPool, RENDER_CONFIG } from '../packages/pdf2img/src/index.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const STATIC_DIR = path.join(__dirname, '../static');
const OUTPUT_DIR = path.join(__dirname, '../output/benchmark');

// 每个文件最多渲染的页数
const MAX_PAGES_TO_RENDER = 10;

/**
 * 动态获取 static 目录下所有 PDF 文件
 */
function getPdfFiles() {
    const files = fs.readdirSync(STATIC_DIR);
    return files
        .filter(f => f.toLowerCase().endsWith('.pdf'))
        .map(name => {
            const filePath = path.join(STATIC_DIR, name);
            const stat = fs.statSync(filePath);
            return {
                name,
                path: filePath,
                size: stat.size,
            };
        })
        .sort((a, b) => a.size - b.size); // 按文件大小排序
}

// 格式化文件大小
function formatSize(bytes) {
    if (bytes < 1024) return `${bytes} B`;
    if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
    return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

// 格式化时间
function formatTime(ms) {
    if (ms < 1000) return `${ms.toFixed(0)} ms`;
    return `${(ms / 1000).toFixed(2)} s`;
}

async function runBenchmark() {
    console.log('='.repeat(70));
    console.log('PDF 转图片性能测试 (PDFium + Sharp)');
    console.log('='.repeat(70));

    // 检查渲染器
    if (!isAvailable()) {
        console.error('❌ 原生渲染器不可用');
        process.exit(1);
    }
    console.log(`渲染器版本: ${getVersion()}`);
    console.log(`默认宽度: ${RENDER_CONFIG.TARGET_RENDER_WIDTH}px`);
    console.log(`最大渲染页数: ${MAX_PAGES_TO_RENDER}`);
    console.log();

    // 动态获取 PDF 文件列表
    const pdfFiles = getPdfFiles();
    console.log(`📁 发现 ${pdfFiles.length} 个 PDF 文件`);
    console.log();

    // 确保输出目录存在
    if (!fs.existsSync(OUTPUT_DIR)) {
        fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }

    const results = [];

    for (const pdfFile of pdfFiles) {
        const pageCount = await getPageCount(pdfFile.path);
        const pagesToRender = Math.min(MAX_PAGES_TO_RENDER, pageCount);

        console.log(`📄 ${pdfFile.name}`);
        console.log(`   大小: ${formatSize(pdfFile.size)}`);
        console.log(`   总页数: ${pageCount}, 渲染: 前 ${pagesToRender} 页`);
        console.log();

        // 测试不同格式
        const formats = ['webp', 'png', 'jpg'];
        const fileResults = { 
            file: pdfFile.name, 
            fileSize: pdfFile.size, 
            pageCount,
            renderedPages: pagesToRender,
            formats: {} 
        };

        for (const format of formats) {
            const outputDir = path.join(OUTPUT_DIR, pdfFile.name.replace('.pdf', ''), format);
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }

            // 预热（仅第一个格式）
            if (format === 'webp') {
                await convert(pdfFile.path, {
                    pages: [1],
                    outputType: 'buffer',
                    format: 'webp',
                });
            }

            const pages = Array.from({ length: pagesToRender }, (_, i) => i + 1);

            const startTime = performance.now();
            const result = await convert(pdfFile.path, {
                pages,
                outputType: 'file',
                outputDir,
                format,
                targetWidth: 1280,
                webp: { quality: 80, method: 4 },
                jpeg: { quality: 85 },
                png: { compressionLevel: 6 },
            });
            const endTime = performance.now();

            const totalTime = endTime - startTime;
            const avgTimePerPage = totalTime / result.renderedPages;

            // 计算输出文件总大小
            const outputFiles = fs.readdirSync(outputDir);
            const totalOutputSize = outputFiles.reduce((sum, f) => {
                return sum + fs.statSync(path.join(outputDir, f)).size;
            }, 0);

            fileResults.formats[format] = {
                totalTime,
                avgTimePerPage,
                outputSize: totalOutputSize,
                pages: result.renderedPages,
            };

            console.log(`   ${format.toUpperCase().padEnd(4)}: ${formatTime(totalTime).padStart(8)} (${formatTime(avgTimePerPage).padStart(6)}/页), 输出 ${formatSize(totalOutputSize).padStart(8)}`);
        }

        results.push(fileResults);
        console.log();
    }

    // 输出汇总表格
    console.log('='.repeat(70));
    console.log(`性能汇总 (前 ${MAX_PAGES_TO_RENDER} 页)`);
    console.log('='.repeat(70));
    console.log();
    console.log('| 文件 | 大小 | 渲染页 | WebP | PNG | JPG |');
    console.log('|------|------|--------|------|-----|-----|');

    for (const r of results) {
        const fileName = r.file.length > 20 ? r.file.slice(0, 17) + '...' : r.file;
        const webp = r.formats.webp ? formatTime(r.formats.webp.totalTime) : '-';
        const png = r.formats.png ? formatTime(r.formats.png.totalTime) : '-';
        const jpg = r.formats.jpg ? formatTime(r.formats.jpg.totalTime) : '-';
        console.log(`| ${fileName.padEnd(20)} | ${formatSize(r.fileSize).padEnd(8)} | ${String(r.renderedPages).padEnd(6)} | ${webp.padEnd(8)} | ${png.padEnd(8)} | ${jpg.padEnd(8)} |`);
    }

    // 获取线程池统计
    const poolStats = getThreadPoolStats();
    
    console.log();
    console.log('📝 说明:');
    console.log('   架构: PDFium 渲染 + Sharp 编码 (piscina 线程池)');
    console.log(`   工作线程: ${poolStats.workers} 个`);
    if (poolStats.initialized) {
        console.log(`   已完成任务: ${poolStats.completed}`);
        console.log(`   线程利用率: ${(poolStats.utilization * 100).toFixed(1)}%`);
    }
    console.log();
    console.log('✅ 性能测试完成');
    console.log(`   输出目录: ${OUTPUT_DIR}`);

    // 销毁线程池
    await destroyThreadPool();

    return results;
}

// 运行测试
runBenchmark().catch(console.error);

File: test/file-benchmark.rs
//! File rendering performance benchmark
//!
//! Tests conversion performance for all PDF files in static directory.
//!
//! # Usage
//!
//! ```bash
//! cargo run --release --bin file-benchmark
//! ```

use pdf2img_core::{convert, get_page_count, is_pdfium_available, get_version, warmup, ConvertOptions, OutputFormat};
use std::fs;
use std::path::PathBuf;
use std::time::Instant;

const MAX_PAGES_TO_RENDER: usize = 10;

struct PdfFile {
    name: String,
    path: PathBuf,
    size: u64,
}

fn get_pdf_files() -> Vec<PdfFile> {
    let static_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("static");

    let mut files: Vec<PdfFile> = fs::read_dir(&static_dir)
        .expect("Failed to read static directory")
        .filter_map(|entry| {
            let entry = entry.ok()?;
            let path = entry.path();
            if path.extension()?.to_str()? == "pdf" {
                let metadata = fs::metadata(&path).ok()?;
                Some(PdfFile {
                    name: path.file_name()?.to_string_lossy().to_string(),
                    path,
                    size: metadata.len(),
                })
            } else {
                None
            }
        })
        .collect();

    files.sort_by_key(|f| f.size);
    files
}

fn format_size(bytes: u64) -> String {
    if bytes < 1024 {
        format!("{} B", bytes)
    } else if bytes < 1024 * 1024 {
        format!("{:.1} KB", bytes as f64 / 1024.0)
    } else {
        format!("{:.1} MB", bytes as f64 / 1024.0 / 1024.0)
    }
}

fn format_time(ms: u32) -> String {
    if ms < 1000 {
        format!("{} ms", ms)
    } else {
        format!("{:.2} s", ms as f64 / 1000.0)
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    println!("{}", "=".repeat(70));
    println!("PDF to Image Performance Benchmark (PDFium)");
    println!("{}", "=".repeat(70));

    // Check renderer
    if !is_pdfium_available() {
        eprintln!("❌ PDFium not available");
        std::process::exit(1);
    }
    println!("Version: {}", get_version());
    println!("Max pages to render: {}", MAX_PAGES_TO_RENDER);

    // Warmup
    print!("Warming up... ");
    let warmup_time = warmup()?;
    println!("done ({} ms)", warmup_time);
    println!();

    // Get PDF files
    let pdf_files = get_pdf_files();
    println!("📁 Found {} PDF files", pdf_files.len());
    println!();

    // Create output directory
    let output_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .unwrap()
        .parent()
        .unwrap()
        .join("output/benchmark");
    fs::create_dir_all(&output_dir)?;

    let formats = vec![OutputFormat::WebP, OutputFormat::Png, OutputFormat::Jpeg];

    for pdf_file in &pdf_files {
        let page_count = get_page_count(&pdf_file.path)?;
        let pages_to_render = MAX_PAGES_TO_RENDER.min(page_count as usize);

        println!("📄 {}", pdf_file.name);
        println!("   Size: {}", format_size(pdf_file.size));
        println!("   Total pages: {}, rendering: first {}", page_count, pages_to_render);

        for format in &formats {
            let pages: Vec<u32> = (1..=pages_to_render as u32).collect();
            
            let options = ConvertOptions {
                format: *format,
                target_width: Some(1280),
                quality: Some(80),
                pages: pages.clone(),
                ..Default::default()
            };

            let start = Instant::now();
            let result = convert(&pdf_file.path.to_string_lossy(), Some(options)).await?;
            let total_time = start.elapsed().as_millis() as u32;

            let total_size: usize = result.pages.iter()
                .filter(|p| p.success)
                .map(|p| p.data.len())
                .sum();

            let avg_time = total_time / result.pages.len() as u32;

            println!(
                "   {:4}: {:>8} ({:>6}/page), output {:>8}",
                match format {
                    OutputFormat::WebP => "WebP",
                    OutputFormat::Png => "PNG",
                    OutputFormat::Jpeg => "JPG",
                },
                format_time(total_time),
                format_time(avg_time),
                format_size(total_size as u64),
            );
        }

        println!();
    }

    println!("{}", "=".repeat(70));
    println!("✅ Benchmark completed");
    println!("   Output directory: {}", output_dir.display());

    Ok(())
}
